client/emv/cmdemv.c

```diff
@@ -0,0 +1,540 @@
+//-----------------------------------------------------------------------------
+// Copyright (C) 2017 Merlok
+//
+// This code is licensed to you under the terms of the GNU GPL, version 2 or,
+// at your option, any later version. See the LICENSE.txt file for the text of
+// the license.
+//-----------------------------------------------------------------------------
+// EMV commands
+//-----------------------------------------------------------------------------
+
+#include "cmdemv.h"
+
+#define APDU_RES_LEN 520
+#define APDU_AID_LEN 50
+
+int UsageCmdHFEMVSelect(void) {
+	PrintAndLog("HELP :  Executes select applet command:\n");
+	PrintAndLog("Usage:  hf emv select [-s][-k][-t] <HEX applet AID>\n");
+	PrintAndLog("  Options:");
+	PrintAndLog("  -s       : select card");
+	PrintAndLog("  -k       : keep field for next command");
+	PrintAndLog("  -t       : TLV decode results\n");
+	PrintAndLog("Samples:");
+	PrintAndLog(" hf emv select -s a00000000101 -> select card, select applet");
+	PrintAndLog(" hf emv select -s -t a00000000101 -> select card, select applet, show result in TLV");
+	return 0;
+}
+
+int CmdHFEMVSelect(const char *cmd) {
+	uint8_t data[APDU_AID_LEN] = {0};
+	int datalen = 0;
+	bool activateField = false;
+	bool leaveSignalON = false;
+	bool decodeTLV = false;
+
+	if (strlen(cmd) < 1) {
+		UsageCmdHFEMVSelect();
+		return 0;
+	}
+	
+	SetAPDULogging(true);
+	
+	int cmdp = 0;
+	while(param_getchar(cmd, cmdp) != 0x00) {
+		char c = param_getchar(cmd, cmdp);
+		if ((c == '-') && (param_getlength(cmd, cmdp) == 2))
+			switch (param_getchar_indx(cmd, 1, cmdp)) {
+				case 'h':
+				case 'H':
+					UsageCmdHFEMVSelect();
+					return 0;
+				case 's':
+				case 'S':
+					activateField = true;
+					break;
+				case 'k':
+				case 'K':
+					leaveSignalON = true;
+					break;
+				case 't':
+				case 'T':
+					decodeTLV = true;
+					break;
+				default:
+					PrintAndLog("Unknown parameter '%c'", param_getchar_indx(cmd, 1, cmdp));
+					return 1;
+		}
+
+		if (isxdigit(c)) {
+			switch(param_gethex_to_eol(cmd, cmdp, data, sizeof(data), &datalen)) {
+			case 1:
+				PrintAndLog("Invalid HEX value.");
+				return 1;
+			case 2:
+				PrintAndLog("AID too large.");
+				return 1;
+			case 3:
+				PrintAndLog("Hex must have even number of digits.");
+				return 1;
+			}
+			
+			// we get all the hex to end of line with spaces
+			break;
+		}
+
+
+		cmdp++;
+	}
+	
```

Comment:
Need to handle `datalen == 0` (no AID)

Reply:
command select can select default application wo AID

Reply:
Indeed.